internal/controller/nicdevice_controller.go

```diff
@@ -0,0 +1,550 @@
+
+type nicDeviceConfigurationStatus struct {
+	device                 *v1alpha1.NicDevice
+	nvConfigUpdateRequired bool
+	rebootRequired         bool
+	lastStageError         error
+}
+
+// Reconcile reconciles the NicConfigurationTemplate object
+func (r *NicDeviceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	devices := &v1alpha1.NicDeviceList{}
+
+	selectorFields := fields.OneTermEqualSelector("status.node", r.nodeName)
+
+	err := r.Client.List(ctx, devices, &client.ListOptions{FieldSelector: selectorFields})
+	if err != nil {
+		log.Log.Error(err, "failed to list NicDevice CRs")
+		return ctrl.Result{}, err
+	}
+
+	if len(devices.Items) == 0 {
+		err = r.maintenanceManager.ReleaseMaintenance(ctx)
+		if err != nil {
+			log.Log.Error(err, "failed to release maintenance")
+			return ctrl.Result{}, err
+		}
+		// Nothing to reconcile
+		return ctrl.Result{}, nil
+	}
+
+	configStatuses := nicDeviceConfigurationStatuses{}
+
+	for _, device := range devices.Items {
+		if device.Spec.Configuration == nil {
+			continue
+		}
+
+		configStatuses = append(configStatuses, &nicDeviceConfigurationStatus{
+			device: &device,
+		})
+	}
+
+	if len(configStatuses) == 0 {
+		err = r.maintenanceManager.ReleaseMaintenance(ctx)
+		if err != nil {
+			log.Log.Error(err, "failed to release maintenance")
+			return ctrl.Result{}, err
+		}
+		// Nothing to reconcile
+		return ctrl.Result{}, nil
+	}
+
+	err = r.handleSpecValidation(ctx, configStatuses)
+	if err != nil {
+		log.Log.Error(err, "failed to validate device's spec")
+		return ctrl.Result{}, err
+	}
+
+	if configStatuses.nvConfigUpdateRequired() {
+		log.Log.V(2).Info("nv config update required, scheduling maintenance")
+
+		err = r.maintenanceManager.ScheduleMaintenance(ctx)
+		if err != nil {
+			log.Log.Error(err, "failed to schedule maintenance for node")
+			return ctrl.Result{}, err
+		}
+
+		maintenanceAllowed, err := r.maintenanceManager.MaintenanceAllowed(ctx)
+		if err != nil {
+			log.Log.Error(err, "failed to get maintenance status")
+			return ctrl.Result{}, err
+		}
+		if !maintenanceAllowed {
+			log.Log.V(2).Info("maintenance not allowed yet, exiting for now")
+			// Maintenance not yet allowed, waiting until then
+			return ctrl.Result{Requeue: true, RequeueAfter: requeueTime}, nil
+		}
+
+		err = r.applyNvConfig(ctx, configStatuses)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	if configStatuses.rebootRequired() {
+		requeue, err := r.handleReboot(ctx, configStatuses)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if requeue {
+			return ctrl.Result{Requeue: true, RequeueAfter: requeueTime}, nil
+		}
+	}
+
+	if !configStatuses.nvConfigReadyForAll() {
+		return ctrl.Result{Requeue: true, RequeueAfter: requeueTime}, nil
+	}
+
+	err = r.applyRuntimeConfig(ctx, configStatuses)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if configStatuses.rebootRequired() {
```

Comment:
It feels a bit weird to have the same `if reboot required` code in the same method. Maybe we can optimize this? But I don't have a better idea but "simply requeue" and ujse the "if reboot required" from above

Reply:
Those are actually different reboots (after nv config apply and runtime config apply) and we don't keep the state in between the calls

Reply:
But this means also that we can be stuck in a reboot loop, right?
If we do the nv config apply, reboot and the runtime config apply fails every time it will nv config apply again, reboot and runtime config apply fails... again. Is this an possible edge case?

Reply:
It works like this:
1. ValidateNvConfig returns whether nvConfig change or reboot are required
2. If nv config change is requested it will apply the new config
3. if the reboot was requested (either before or after nv config change) it will reboot the node
4. The second time the nv config is already applied so the reboot will not be requested from this step.
5. If the runtime config failed to apply it will retry again. Reboot is only required if we've had some previously setup configuration that does not match the spec (we save that as lastAppliedState annotation in the CR)

There potentially is a reboot loop when nv config was not properly applied after the reboot and we are looking into whether it's even possible on a healthy device